File: app/src/main/java/com/bitcoin/merchant/app/screens/TransactionsHistoryFragment.java
package com.bitcoin.merchant.app.screens;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.BroadcastReceiver;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.ContentValues;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Typeface;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.style.RelativeSizeSpan;
import android.text.style.StyleSpan;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;

import androidx.fragment.app.Fragment;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.bitcoin.merchant.app.MainActivity;
import com.bitcoin.merchant.app.R;
import com.bitcoin.merchant.app.database.DBControllerV3;
import com.bitcoin.merchant.app.database.PaymentRecord;
import com.bitcoin.merchant.app.util.DateUtil;
import com.bitcoin.merchant.app.util.MonetaryUtil;
import com.bitcoin.merchant.app.util.PrefsUtil;
import com.crashlytics.android.Crashlytics;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import static android.content.Context.CLIPBOARD_SERVICE;

public class TransactionsHistoryFragment extends Fragment {
    private static final String TAG = "TransactionsHistory";
    private TransactionAdapter adapter = null;
    private ListView listView = null;
    private LinearLayout noTxHistoryLv = null;
    private SwipeRefreshLayout swipeLayout = null;
    private Activity thisActivity = null;
    private volatile boolean ready;
    private final BroadcastReceiver fragmentBroadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (MainActivity.ACTION_QUERY_ALL_UXTO_FINISHED.equals(intent.getAction())) {
                swipeLayout.setRefreshing(false);
            }
        }
    };

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        ready = true;
        View rootView = inflater.inflate(getResources().getLayout(R.layout.fragment_transaction), container, false);
        initListView(rootView);
        swipeLayout = rootView.findViewById(R.id.swipe_container);
        swipeLayout.setProgressViewEndTarget(false, (int) (getResources().getDisplayMetrics().density * (72 + 20)));
        swipeLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                new LoadTxFromDatabaseTask(true).execute();
            }
        });
        swipeLayout.setColorSchemeResources(
                R.color.bitcoindotcom_darker_green,
                R.color.bitcoindotcom_green,
                R.color.bitcoindotcom_darkest_green);
        thisActivity = getActivity();
        IntentFilter filter = new IntentFilter();
        filter.addAction(MainActivity.ACTION_QUERY_ALL_UXTO_FINISHED);
        LocalBroadcastManager.getInstance(thisActivity).registerReceiver(fragmentBroadcastReceiver, filter);
        return rootView;
    }

    @Override
    public void onDestroyView() {
        ready = false;
        super.onDestroyView();
        LocalBroadcastManager.getInstance(thisActivity).unregisterReceiver(fragmentBroadcastReceiver);
    }

    private void initListView(View rootView) {
        adapter = new TransactionAdapter();
        listView = rootView.findViewById(R.id.txList);
        noTxHistoryLv = rootView.findViewById(R.id.no_tx_history_lv);
        listView.setAdapter(adapter);
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                showTransactionMenu(id);
            }
        });
        listView.setOnScrollListener(new AbsListView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(AbsListView view, int scrollState) {
            }

            @Override
            public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
                if (listView.getChildAt(0) != null) {
                    swipeLayout.setEnabled(listView.getFirstVisiblePosition() == 0 && listView.getChildAt(0).getTop() == 0);
                }
            }
        });
        if (adapter.getCount() == 0) {
            listView.setVisibility(View.GONE);
            noTxHistoryLv.setVisibility(View.VISIBLE);
        } else {
            listView.setVisibility(View.VISIBLE);
            noTxHistoryLv.setVisibility(View.GONE);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        new LoadTxFromDatabaseTask(false).execute();
    }

    private void showTransactionMenu(final long item) {
        final ContentValues val = adapter.mListItems.get((int) item);
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        final String tx = val.getAsString("tx");
        final String address = val.getAsString("iad");
        builder.setTitle(tx);
        builder.setIcon(R.drawable.ic_launcher);
        builder.setItems(new CharSequence[]{
                        "View transaction",
                        "View all TX with this address",
                        "Copy transaction",
                        "Copy address",
                },
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                        switch (which) {
                            case 0: {
                                openExplorer("https://explorer.bitcoin.com/bch/tx/" + tx);
                                break;
                            }
                            case 1: {
                                openExplorer("https://explorer.bitcoin.com/bch/address/" + address);
                                break;
                            }
                            case 2: {
                                copyToClipboard(tx);
                                break;
                            }
                            case 3: {
                                copyToClipboard(address);
                                break;
                            }
                        }
                    }
                });
        builder.create().show();
    }

    private void openExplorer(String uri) {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
        startActivity(intent);
    }

    private void copyToClipboard(String text) {
        final ClipboardManager clipboardManager = (ClipboardManager) getContext().getSystemService(CLIPBOARD_SERVICE);
        if (clipboardManager != null) {
            ClipData clipData = ClipData.newPlainText("Source Text", text);
            clipboardManager.setPrimaryClip(clipData);
        }
    }

    protected boolean isSafe() {
        if (!ready || thisActivity == null || adapter == null) {
            return false; // view not yet created
        }
        return !(this.isRemoving() || this.getActivity() == null || this.isDetached() || !this.isAdded() || this.getView() == null);
    }

    public void addTx(final ContentValues val) {
        try {
            if (!isSafe()) {
                return;
            }
            listView.setVisibility(View.VISIBLE);
            noTxHistoryLv.setVisibility(View.GONE);
            thisActivity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    if (!isSafe()) {
                        return;
                    }
                    adapter.addNewPayment(val);
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "addTx", e);
            Crashlytics.logException(e);
        }
    }

    public void updateTx(final ContentValues val) {
        try {
            if (!isSafe()) {
                return;
            }
            listView.setVisibility(View.VISIBLE);
            noTxHistoryLv.setVisibility(View.GONE);
            thisActivity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    new LoadTxFromDatabaseTask(false).execute();
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "updateTx", e);
            Crashlytics.logException(e);
        }
    }

    private class LoadTxFromDatabaseTask extends AsyncTask<Void, Void, ArrayList<ContentValues>> {
        private boolean queryServer;

        public LoadTxFromDatabaseTask(boolean queryServer) {
            this.queryServer = queryServer;
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            swipeLayout.setRefreshing(true);
        }

        @Override
        protected ArrayList<ContentValues> doInBackground(Void... params) {
            if (!ready) {
                return null;
            }
            String address = PrefsUtil.getInstance(getActivity()).getValue(PrefsUtil.MERCHANT_KEY_MERCHANT_RECEIVER, "");
            if ((address != null) && (address.length() > 0)) {
                try {
                    return new DBControllerV3(getActivity()).getAllPayments();
                } catch (Exception e) {
                    Log.e(TAG, "getAllPayments", e);
                    Crashlytics.logException(e);
                }
            }
            return null;
        }

        @Override
        protected void onPostExecute(ArrayList<ContentValues> result) {
            super.onPostExecute(result);
            if (ready) {
                if (result != null && adapter != null) {
                    adapter.reset(result);
                    if (result.size() != 0) {
                        listView.setVisibility(View.VISIBLE);
                        noTxHistoryLv.setVisibility(View.GONE);
                    }
                }
                if (queryServer) {
                    findAllPotentialMissingTx();
                } else {
                    if (result != null && result.size() != 0) {
                        listView.setVisibility(View.VISIBLE);
                        noTxHistoryLv.setVisibility(View.GONE);
                    }
                    swipeLayout.setRefreshing(false);
                }
            }
        }
    }

    private void findAllPotentialMissingTx() {
        LocalBroadcastManager.getInstance(thisActivity).sendBroadcast(new Intent(MainActivity.ACTION_QUERY_ALL_UXTO));
    }

    private class TransactionAdapter extends BaseAdapter {
        private final List<ContentValues> mListItems = new ArrayList<>();
        private final LayoutInflater inflater;

        TransactionAdapter() {
            inflater = (LayoutInflater) getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        }

        public void addNewPayment(ContentValues val) {
            for (ContentValues c : mListItems) {
                if (val.getAsString("tx").equals(c.getAsString("tx"))) {
                    return; // already added
                }
            }
            mListItems.add(0, val);
            notifyDataSetChanged();
        }

        public void reset(ArrayList<ContentValues> vals) {
            mListItems.clear();
            mListItems.addAll(vals);
            notifyDataSetChanged();
        }

        @Override
        public int getCount() {
            return mListItems.size();
        }

        @Override
        public String getItem(int position) {
            return mListItems.get(position).getAsString("iad");
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            View view = convertView != null ? convertView
                    : inflater.inflate(R.layout.list_item_transaction, parent, false);
            try {
                ContentValues vals = mListItems.get(position);
                PaymentRecord r = new PaymentRecord(vals);
                setupView(view, r.bchAmount, r.fiatAmount, r.timeInSec, r.confirmations);
            } catch (Exception e) {
                Log.e(TAG, "getView", e);
            }
            return view;
        }

        private void setupView(View view, Long bch, String fiat, long timeInSec, int confirmations) {
            String date = DateUtil.getInstance().format(timeInSec);
            SpannableStringBuilder ds = new SpannableStringBuilder(date);
            int idx = date.indexOf("@");
            if (idx != -1) {
                ds.setSpan(new StyleSpan(Typeface.NORMAL), 0, idx, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                ds.setSpan(new RelativeSizeSpan(0.75f), idx, date.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            }
            TextView tvDate = view.findViewById(R.id.tv_date);
            tvDate.setText(ds);
            tvDate.setAlpha(0.7f);
            // display coin amount
            long amount = Math.abs(bch);
            SpannableStringBuilder coinSpan = new SpannableStringBuilder(thisActivity.getResources().getString(R.string.bitcoin_currency_symbol));
            coinSpan.setSpan(new RelativeSizeSpan((float) 0.75), 0, 1, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            TextView coinView = view.findViewById(R.id.tv_amount_coin);
            String displayValue = MonetaryUtil.getInstance(thisActivity).getDisplayAmountWithFormatting(amount);
            coinView.setText(displayValue + " " + coinSpan);
            // display fiat amount
            TextView fiatView = view.findViewById(R.id.tv_amount_fiat);
            if (fiat != null && fiat.length() > 1) {
                SpannableStringBuilder fiatSpan = new SpannableStringBuilder(fiat.subSequence(0, 1));
                fiatSpan.setSpan(new RelativeSizeSpan((float) 0.75), 0, 1, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                fiatView.setText(fiatSpan + " " + fiat.substring(1));
            } else {
                fiatView.setText("");
            }
            float secs = System.currentTimeMillis() / 1000f;
            if ((secs - timeInSec) > TimeUnit.HOURS.toSeconds(3)) {
                confirmations = 2;
            }
            ImageView ivStatus = view.findViewById(R.id.iv_status);
            ivStatus.setImageResource(getIcon(bch, confirmations));
            ivStatus.setAlpha(getAlpha(bch, confirmations));
        }

        private int getIcon(Long bch, int confirmations) {
            if (bch < 0L) { // under/over payment
                return R.drawable.ic_warning_black_18dp;
            } else if (confirmations <= 0) {
                return R.drawable.ic_done_white_24dp;
            } else if (confirmations == 1) {
                return R.drawable.ic_done_white_24dp;
            } else { // 2 or more confirmations
                return R.drawable.ic_doublecheck_white_24dp;
            }
        }

        private float getAlpha(Long bch, int confirmations) {
            if (bch < 0L) { // under/over payment
                return 1.0f;
            } else if (confirmations <= 0) {
                return 0f;
            } else {
                return 1.0f;
            }
        }
    }
}

File: app/src/main/java/com/bitcoin/merchant/app/screens/TransactionsHistoryFragment.kt
package com.bitcoin.merchant.app.screens

import android.app.AlertDialog
import android.content.*
import android.graphics.Color
import android.graphics.Typeface
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.style.RelativeSizeSpan
import android.text.style.StyleSpan
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.content.ContextCompat
import androidx.lifecycle.lifecycleScope
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.bitcoin.merchant.app.R
import com.bitcoin.merchant.app.database.PaymentRecord
import com.bitcoin.merchant.app.database.toPaymentRecord
import com.bitcoin.merchant.app.model.Analytics
import com.bitcoin.merchant.app.screens.features.ToolbarAwareFragment
import com.bitcoin.merchant.app.util.DateUtil
import com.bitcoin.merchant.app.util.MonetaryUtil
import com.bitcoin.merchant.app.util.PrintUtil
import com.bitcoin.merchant.app.util.Settings
import com.google.android.material.floatingactionbutton.FloatingActionButton
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.bitcoindotcom.bchprocessor.bip70.model.Bip70Action
import java.util.*
import java.util.concurrent.TimeUnit
import kotlin.collections.ArrayList
import kotlin.math.abs

class TransactionsHistoryFragment : ToolbarAwareFragment() {
    private lateinit var adapter: TransactionAdapter
    private lateinit var txListArea: ViewGroup
    private lateinit var listView: ListView
    private lateinit var printButton: FloatingActionButton
    private lateinit var noTxHistoryLv: LinearLayout
    private lateinit var dateFilterArea: ViewGroup
    private lateinit var startDateButton: Button
    private lateinit var endDateButton: Button
    private lateinit var datePresetArea: ViewGroup
    private lateinit var todayButton: Button
    private lateinit var yesterdayButton: Button
    private lateinit var allButton: Button
    private lateinit var datePicker: DatePicker
    private lateinit var timePicker: TimePicker
    private lateinit var dateSelectionArea: ViewGroup
    private lateinit var dateTimeClearButton: Button
    private lateinit var dateTimeSaveButton: Button
    private lateinit var swipeLayout: SwipeRefreshLayout
    private val fragmentBroadcastReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (Bip70Action.QUERY_ALL_TX_FROM_BITCOIN_COM_PAY == intent.action) {
                swipeLayout.isRefreshing = false
            }
        }
    }
    @Volatile
    private var ready = false
    private var startDateTime: Long = 0
    private var endDateTime: Long = 0

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        ready = true
        val rootView = inflater.inflate(resources.getLayout(R.layout.fragment_transaction), container, false)
        initListView(rootView)
        setToolbarAsBackButton()
        setToolbarTitle(R.string.menu_transactions)
        swipeLayout = rootView.findViewById(R.id.swipe_container)
        swipeLayout.setProgressViewEndTarget(false, (resources.displayMetrics.density * (72 + 20)).toInt())
        swipeLayout.setOnRefreshListener { loadTxFromDatabaseTask(true) }
        swipeLayout.setColorSchemeResources(
                R.color.bitcoindotcom_darker_green,
                R.color.bitcoindotcom_green,
                R.color.bitcoindotcom_darkest_green)
        val filter = IntentFilter()
        filter.addAction(Bip70Action.QUERY_ALL_TX_FROM_BITCOIN_COM_PAY)
        LocalBroadcastManager.getInstance(activity).registerReceiver(fragmentBroadcastReceiver, filter)
        return rootView
    }

    override fun onDestroyView() {
        ready = false
        super.onDestroyView()
        LocalBroadcastManager.getInstance(activity).unregisterReceiver(fragmentBroadcastReceiver)
    }

    private fun initListView(rootView: View) {
        // date filters
        dateFilterArea = rootView.findViewById(R.id.dateFilterArea)
        startDateButton = rootView.findViewById(R.id.startDateButton)
        endDateButton = rootView.findViewById(R.id.endDateButton)
        // date presets
        datePresetArea = rootView.findViewById(R.id.datePresetArea)
        todayButton = rootView.findViewById(R.id.todayButton)
        yesterdayButton = rootView.findViewById(R.id.yesterdayButton)
        allButton = rootView.findViewById(R.id.allButton)
        // date selection
        dateSelectionArea = rootView.findViewById(R.id.dateSelectionArea)
        datePicker = rootView.findViewById(R.id.datePicker)
        timePicker = rootView.findViewById(R.id.timePicker)
        dateTimeClearButton = rootView.findViewById(R.id.dateTimeClearButton)
        dateTimeSaveButton = rootView.findViewById(R.id.dateTimeSaveButton)
        // transactions
        adapter = TransactionAdapter()
        txListArea = rootView.findViewById(R.id.txListArea)
        listView = rootView.findViewById(R.id.txList)
        printButton = rootView.findViewById(R.id.printButton)
        noTxHistoryLv = rootView.findViewById(R.id.no_tx_history_lv)
        listView.adapter = adapter
        listView.setOnItemClickListener { _, _, _, id -> showTransactionMenu(id) }
        listView.setOnScrollListener(object : AbsListView.OnScrollListener {
            override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {}
            override fun onScroll(view: AbsListView, firstVisibleItem: Int, visibleItemCount: Int, totalItemCount: Int) {
                listView.getChildAt(0)?.also {
                    swipeLayout.isEnabled = listView.firstVisiblePosition == 0 && it.top == 0
                }
            }
        })
        setTxListVisibility(adapter.count > 0)
        startDateButton.setOnClickListener {
            getDateTime(0, 0, 0, startDateTime, object : DateTimeCallback {
                override fun invoke(timeInMillis: Long) {
                    enterDateRange(null, timeInMillis, endDateTime)
                }
            })
        }
        endDateButton.setOnClickListener {
            getDateTime(23, 59, 59, endDateTime, object : DateTimeCallback {
                override fun invoke(timeInMillis: Long) {
                    enterDateRange(null, startDateTime, timeInMillis)
                }
            })
        }
        todayButton.setOnClickListener {
            val time = Calendar.getInstance() // today
            time.set(Calendar.HOUR_OF_DAY, 0)
            time.set(Calendar.MINUTE, 0)
            time.set(Calendar.SECOND, 0)
            val startTime = time.timeInMillis
            time.set(Calendar.HOUR_OF_DAY, 23)
            time.set(Calendar.MINUTE, 59)
            time.set(Calendar.SECOND, 59)
            val endTime = time.timeInMillis
            enterDateRange(todayButton, startTime, endTime)
        }
        yesterdayButton.setOnClickListener {
            val time = Calendar.getInstance() // today
            time.add(Calendar.DAY_OF_MONTH, -1) // yesterday
            time.set(Calendar.HOUR_OF_DAY, 0)
            time.set(Calendar.MINUTE, 0)
            time.set(Calendar.SECOND, 0)
            val startTime = time.timeInMillis
            time.set(Calendar.HOUR_OF_DAY, 23)
            time.set(Calendar.MINUTE, 59)
            time.set(Calendar.SECOND, 59)
            val endTime = time.timeInMillis
            enterDateRange(yesterdayButton, startTime, endTime)
        }
        allButton.setOnClickListener {
            enterDateRange(allButton, 0L, 0L)
        }
        printButton.setOnClickListener {
            val reportHtml = PrintUtil.createReportHtml(activity, startDateTime, endDateTime, adapter.mFilteredItems)
            PrintUtil.printHtml(activity, reportHtml)
        }
    }

    private fun enterDateRange(presetButton: Button?, newStartTime: Long, newEndTime: Long) {
        setSelectedDatePreset(presetButton)

        startDateTime = newStartTime
        val startString = if (newStartTime == 0L) getString(R.string.empty_date_string) else DateUtil.instance.format(newStartTime)
        startDateButton.text = startString

        endDateTime = newEndTime
        val endString = if (newEndTime == 0L) getString(R.string.empty_date_string) else DateUtil.instance.format(newEndTime)
        endDateButton.text = endString

        dateSelectionArea.visibility = View.GONE

        adapter.filter()
    }

    private fun setSelectedDatePreset(selectedButton: Button?) {
        val white = ContextCompat.getColor(requireContext(), R.color.white)
        val green = ContextCompat.getColor(requireContext(), R.color.bitcoindotcom_green)
        val black = Color.BLACK

        todayButton.setBackgroundColor(white)
        todayButton.setTextColor(black)
        yesterdayButton.setBackgroundColor(white)
        yesterdayButton.setTextColor(black)
        allButton.setBackgroundColor(white)
        allButton.setTextColor(black)

        if (selectedButton != null) {
            selectedButton.setBackgroundColor(green)
            selectedButton.setTextColor(white)
        }
    }

    private fun setTxListVisibility(enabled: Boolean) {
        val listVisibility = if (enabled) View.VISIBLE else View.GONE
        txListArea.visibility = listVisibility
        dateFilterArea.visibility = listVisibility
        datePresetArea.visibility = listVisibility
        dateSelectionArea.visibility = View.GONE
        noTxHistoryLv.visibility = if (enabled) View.GONE else View.VISIBLE
    }

    override fun onResume() {
        super.onResume()
        loadTxFromDatabaseTask(false)
    }

    private fun showTransactionMenu(item: Long) {
        val paymentRecord = adapter.mListItems[item.toInt()].toPaymentRecord()
        val txId = paymentRecord.tx ?: ""
        val emojiLink = String(Character.toChars(0x1F517))
        val emojiReceipt = String(Character.toChars(0x1F9FE))
        val emojiClipboard = String(Character.toChars(0x1F4CB))
        val builder = AlertDialog.Builder(activity)
        builder.setTitle("${paymentRecord.fiatAmount} - ${Date(paymentRecord.timeInSec*1000)}")
        builder.setIcon(R.mipmap.ic_launcher)
        builder.setItems(arrayOf<CharSequence>(
                "$emojiLink ${getString(R.string.inspect_tx_link_view_transaction)}",
                "$emojiClipboard ${getString(R.string.inspect_tx_copy_transaction)}",
                "$emojiReceipt ${getString(R.string.inspect_tx_print_receipt)}")
        ) { dialog: DialogInterface, which: Int ->
            dialog.dismiss()
            when (which) {
                0 -> Analytics.tx_id_explorer_launched.send()
                1 -> Analytics.tx_id_copied.send()
            }
            when (which) {
                0 -> openExplorer(getString(R.string.url_explorer_bitcoin_com) + "/bch/tx/$txId")
                1 -> copyToClipboard(txId)
                2 -> printReceipt(paymentRecord)
            }
        }
        builder.create().show()
    }

    private fun printReceipt(paymentRecord: PaymentRecord) {
        val receiptHtml = PrintUtil.createReceiptHtml(activity, paymentRecord)
        PrintUtil.printHtml(activity, receiptHtml)
    }

    private fun openExplorer(uri: String) {
        startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(uri)))
    }

    private fun copyToClipboard(text: String) {
        val clipboardManager = activity.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val clipData = ClipData.newPlainText("Source Text", text)
        clipboardManager.setPrimaryClip(clipData)
    }

    private interface DateTimeCallback {
        fun invoke(timeInMillis: Long)
    }

    private fun getDateTime(initialHour: Int, initialMinute: Int, seconds: Int, currentTime: Long, callback: DateTimeCallback) {
        val now = Calendar.getInstance()
        var defaultYear = now.get(Calendar.YEAR)
        var defaultMonth = now.get(Calendar.MONTH)
        var defaultDay = now.get(Calendar.DAY_OF_MONTH)
        var defaultHour = initialHour
        var defaultMinute = initialMinute
        if (currentTime != 0L) {
            val calendar = Calendar.getInstance()
            calendar.setTime(Date(currentTime))
            defaultYear = calendar.get(Calendar.YEAR)
            defaultMonth = calendar.get(Calendar.MONTH)
            defaultDay = calendar.get(Calendar.DAY_OF_MONTH)
            defaultHour = calendar.get(Calendar.HOUR)
            defaultMinute = calendar.get(Calendar.MINUTE)
        }

        datePicker.visibility = View.VISIBLE
        timePicker.visibility = View.GONE
        dateSelectionArea.visibility = View.VISIBLE

        datePicker.init(defaultYear, defaultMonth, defaultDay, null)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            timePicker.hour = defaultHour
            timePicker.minute = defaultMinute
        }
        else {
            timePicker.currentHour = defaultHour
            timePicker.currentMinute = defaultMinute
        }

        dateTimeSaveButton.setOnClickListener {
            // switch from date to time
            timePicker.visibility = View.VISIBLE
            datePicker.visibility = View.GONE
            dateTimeSaveButton.setOnClickListener {
                val newTime = Calendar.getInstance()
                newTime.set(Calendar.YEAR, datePicker.year)
                newTime.set(Calendar.MONTH, datePicker.month)
                newTime.set(Calendar.DAY_OF_MONTH, datePicker.dayOfMonth)
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    newTime.set(Calendar.HOUR_OF_DAY, timePicker.hour)
                    newTime.set(Calendar.MINUTE, timePicker.minute)
                }
                else {
                    newTime.set(Calendar.HOUR_OF_DAY, timePicker.currentHour)
                    newTime.set(Calendar.MINUTE, timePicker.currentMinute)
                }
                newTime.set(Calendar.SECOND, seconds)
                dateSelectionArea.visibility = View.GONE
                callback.invoke(newTime.timeInMillis)
            }
        }
        dateTimeClearButton.setOnClickListener {
            dateSelectionArea.visibility = View.GONE
            callback.invoke(0L)
        }
    }

    // view not yet created
    protected val isSafe: Boolean
        get() = ready && !this.isRemoving && !this.isDetached && this.isAdded && this.view != null

    private fun requestToDownloadAllTx() { // TODO use merchant server to query all TX
        LocalBroadcastManager.getInstance(activity).sendBroadcast(Intent(Bip70Action.QUERY_ALL_TX_FROM_BITCOIN_COM_PAY))
    }

    private fun loadTxFromDatabaseTask(queryServer: Boolean) {
        if (!isSafe || !Settings.getPaymentTarget(activity).isValid) return
        viewLifecycleOwner.lifecycleScope.launch {
            swipeLayout.isRefreshing = true
            val txs: ArrayList<ContentValues>? = withContext(Dispatchers.IO) {
                try {
                    app.db.allPayments
                } catch (e: Exception) {
                    Analytics.error_db_read_tx.sendError(e)
                    Log.e(TAG, "getAllPayments", e)
                    null
                }
            }
            swipeLayout.isRefreshing = false
            if (!isSafe) return@launch
            txs?.also {
                adapter.reset(txs)
                if (txs.size != 0) {
                    setTxListVisibility(true)
                }
            }
            if (queryServer) {
                requestToDownloadAllTx()
            }
        }
    }

    private inner class TransactionAdapter internal constructor() : BaseAdapter() {
        val mListItems: MutableList<ContentValues> = ArrayList()
        val mFilteredItems: MutableList<ContentValues> = ArrayList()
        private val inflater: LayoutInflater = activity.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

        fun reset(values: ArrayList<ContentValues>) {
            mListItems.clear()
            mListItems.addAll(values)

            filterByDateRange()
            notifyDataSetChanged()
        }

        fun filter() {
            filterByDateRange()
            notifyDataSetChanged()
        }

        private fun filterByDateRange() {
            mFilteredItems.clear()
            for (item in mListItems) {
                val timeInSec = item.toPaymentRecord().timeInSec
                val timeInMillis = TimeUnit.SECONDS.toMillis(timeInSec)
                val shouldHide = (startDateTime != 0L && timeInMillis < startDateTime) || (endDateTime != 0L && timeInMillis > endDateTime)
                if (!shouldHide) {
                    mFilteredItems.add(item)
                }
            }
        }

        override fun getCount(): Int {
            return mFilteredItems.size
        }

        override fun getItem(position: Int): String {
            return mFilteredItems[position].getAsString("iad")
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val view = convertView
                    ?: inflater.inflate(R.layout.list_item_transaction, parent, false)
            try {
                val r = mFilteredItems[position].toPaymentRecord()
                setupView(view, r.bchAmount, r.fiatAmount, r.timeInSec, r.confirmations)
            } catch (e: Exception) {
                Analytics.error_rendering.sendError(e)
                Log.e(TAG, "getView", e)
            }
            return view
        }

        private fun setupView(view: View, bch: Long, fiat: String?, timeInSec: Long, confirmations: Int) {
            var confirmations = confirmations
            val date: String = DateUtil.instance.format(TimeUnit.SECONDS.toMillis(timeInSec))
            val ds = SpannableStringBuilder(date)
            val idx = date.indexOf("@")
            if (idx != -1) {
                ds.setSpan(StyleSpan(Typeface.NORMAL), 0, idx, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                ds.setSpan(RelativeSizeSpan(0.75f), idx, date.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            val tvDate = view.findViewById<TextView>(R.id.tv_date)
            tvDate.text = ds
            tvDate.alpha = 0.7f
            // display coin amount
            val amount = abs(bch)
            val coinSpan = SpannableStringBuilder("BCH")
            coinSpan.setSpan(RelativeSizeSpan(0.75.toFloat()), 0, 1, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            val coinView = view.findViewById<TextView>(R.id.tv_amount_coin)
            val displayValue = MonetaryUtil.instance.getDisplayAmountWithFormatting(amount)
            coinView.text = "$displayValue $coinSpan"
            // display fiat amount
            val fiatView = view.findViewById<TextView>(R.id.tv_amount_fiat)
            fiatView.text = if (fiat != null && fiat.length > 1) {
                val fiatSpan = SpannableStringBuilder(fiat.subSequence(0, 1))
                fiatSpan.setSpan(RelativeSizeSpan(0.75.toFloat()), 0, 1, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                "$fiatSpan ${fiat.substring(1)}"
            } else {
                ""
            }
            val secs = System.currentTimeMillis() / 1000f
            if (secs - timeInSec > TimeUnit.HOURS.toSeconds(3)) {
                confirmations = 2
            }
            val ivStatus = view.findViewById<ImageView>(R.id.iv_status)
            ivStatus.setImageResource(getIcon(bch, confirmations))
            ivStatus.alpha = getAlpha(bch, confirmations)
        }

        private fun getIcon(bch: Long, confirmations: Int): Int {
            return when {
                confirmations < 0 -> R.drawable.ic_warning_black_18dp // under/over payment, legacy: can't happen anymore with BIP-70
                confirmations in 0..1 -> R.drawable.ic_done_white_24dp
                else -> R.drawable.ic_doublecheck_white_24dp // 2 or more confirmations
            }
        }

        private fun getAlpha(bch: Long, confirmations: Int): Float {
            return when {
                confirmations < 0 -> 1.0f // under/over payment, legacy: can't happen anymore with BIP-70
                confirmations in 0..1 -> 0f
                else -> 1.0f
            }
        }
    }

    override fun canFragmentBeDiscardedWhenInBackground(): Boolean {
        return true
    }

    companion object {
        private const val TAG = "BCR-TransactionsHistory"
    }
}